usdai/main.py

```diff
@@ -0,0 +1,135 @@
+                if configs:
+                    # Mint ratio is scaled by 1e4 (e.g. 9950 = 99.5%)
+                    mint_ratio_raw = int(configs[0].get("value", 10000))
+                    mint_ratio = mint_ratio_raw
+                
+        except Exception as e:
+            print(f"API Error: {e}")
+
+        # Derived Collateral from Mint Ratio
+        # Scaling: mint_ratio is in bps (1e4). 
+        # So mint_ratio_fmt = mint_ratio / 10000.
+        mint_ratio_fmt = mint_ratio / 10000
+        
+        # Avoid division by zero
+        required_collateral = 0
+        if mint_ratio_fmt > 0:
+            required_collateral = usdai_supply_fmt / mint_ratio_fmt
+
+        print(f"\n--- USDai Stats ---")
+        print(f"USDai Supply:    ${usdai_supply_fmt:,.2f}")
+        print(f"Mint Ratio:      {mint_ratio_fmt:.4f}")
+        
+        collateral_metric = required_collateral
+        # Buffer = Collateral - Supply
+        buffer = collateral_metric - usdai_supply_fmt
+        
+        print(f"Collateral:      ${collateral_metric:,.2f}")
+        print(f"Buffer:          ${buffer:,.2f}")
+        
+        if collateral_metric > 0:
+            # 1. Check for Mint Ratio Change (Critical)
+            cache_key_ratio = f"{PROTOCOL}_mint_ratio"
+            last_ratio_str = get_last_value_for_key_from_file(cache_filename, cache_key_ratio)
+            last_ratio = int(last_ratio_str) if last_ratio_str != 0 else 0
+            
+            if last_ratio != 0 and last_ratio != mint_ratio:
+                msg = f"⚠️ *USDai Mint Ratio Changed*\n\nOld: {last_ratio/10000:.4f}\nNew: {mint_ratio/10000:.4f}"
+                send_telegram_message(msg, PROTOCOL)
+            
+            # Always update ratio cache
+            write_last_value_to_file(cache_filename, cache_key_ratio, mint_ratio)
+
+            # 2. Check for Buffer Drop (Activity)
+            cache_key_buffer = f"{PROTOCOL}_buffer"
+            last_buffer_str = get_last_value_for_key_from_file(cache_filename, cache_key_buffer)
+            last_buffer = float(last_buffer_str) if last_buffer_str != 0 else 0.0
+
+            if last_buffer == 0:
+                write_last_value_to_file(cache_filename, cache_key_buffer, buffer)
+            else:
+                change = buffer - last_buffer
+                # Alert if buffer drops by more than $10,000 (approx $2M redemption)
+                if change < -10000: 
+                    msg = f"📉 *USDai Buffer Drop Alert*\n\nBuffer dropped by ${abs(change):,.2f}!\nOld Buffer: ${last_buffer:,.2f}\nNew Buffer: ${buffer:,.2f}\n(Collateral: ${collateral_metric:,.2f})"
+                    send_telegram_message(msg, PROTOCOL)
+                    write_last_value_to_file(cache_filename, cache_key_buffer, buffer)
+                elif change > 0:
+                     # Update baseline on rise
+                     write_last_value_to_file(cache_filename, cache_key_buffer, buffer)
```

Comment:
The function `write_last_value_to_file` is called in every path here. Extract it from if else block, so you call it at the end. This will let you simplify first if/else block. Less code is always better :)